proposals/NNNN-typed-throws-concurrency-library.md

```diff
@@ -0,0 +1,319 @@
+
+The [`Task`](https://developer.apple.com/documentation/swift/task) APIs have a
+`Failure` type similarly to `Result`, but use a pattern of overloading on
+`Failure == any Error` and `Failure == Never` to handling throwing and
+non-throwing versions. For example, the `Task` initializer is defined as the
+following overloaded pair:
+
+```swift
+init(priority: TaskPriority?, operation: () async -> Success) where Failure == Never
+init(priority: TaskPriority?, operation: () async throws -> Success) where Failure == any Error
+```
+
+These two initializers can be replaced with a single initializer using typed throws:
+
+```swift
+init(priority: TaskPriority?, operation: () async throws(Failure) -> Success)
+```
+
+The result is both more expressive (maintaining typed error information) and
+simpler (because a single initializer suffices). The same transformation can be
+applied to the `detached` function that creates detached tasks, where the two
+overloads are replaced with the following:
+
+```swift
+@discardableResult
+static func detached(
+    priority: TaskPriority? = nil,
+    operation: @escaping () async throws(Failure) -> Success
+) -> Task<Success, Failure>
+```
+
+Finally, the `value` property of `Task` is similarly overloaded:
+
+```swift
+extension Task where Failure == Never {}
+  var value: Success { get async }
+}
+extension Task where Failure == any Error {
+  var value: Success { get async throws }
+}
+```
+
+These two can be replaced with a single property:
+
+```swift
+var value: Success { get async throws(Failure) }
+```
+
+### Continuations
+
+Currently, there are two variants of each continuation creation method to
+accommodate for the throwing and non-throwing variant. Since, the natural
+spelling would be without the `Throwing` in the method name. We propose
+to add the following two new methods:
+
+```swift
+public func withCheckedContinuation<T, Failure: Error>(
+    function: String = #function,
+    _ body: (CheckedContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+
+public func withUnsafeContinuation<T, Failure: Error>(
+  _ fn: (UnsafeContinuation<T, Failure>) -> Void
+) async throws(Failure) -> T
+```
+
+#### Task cancellation
+
+A few of the `Task` APIs are documented to only throw `CancellationError` and
+can adopt typed throws. For example, `checkCancellation`:
+
+```swift
+public static func checkCancellation() throws(CancellationError)
+```
+
+Similarly, the `sleep` APIs will only throw on cancellation:
+
+```swift
+public static func sleep<C: Clock>(
+  until deadline: C.Instant,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+
+public static func sleep<C: Clock>(
+  for duration: C.Instant.Duration,
+  tolerance: C.Instant.Duration? = nil,
+  clock: C = ContinuousClock()
+) async throws(CancellationError)
+```
+
+Additionally, the `ContinuousClock` and `SuspendingClock` `sleep` methods will also
+adopt typed throws:
```

Comment:
are we sure about this? It'll be a source break potentially; since those are protocol requirements and users "should have"™ only thrown cancellation error but in practice they may have thrown anything...

Since we're Swift6 planning here we can accept such breaks I suppose -- is that the thinking here? We should spell that out.

---

**Franz:** Good point. I think we actually can't adopt a typed throw here.